File: MATLAB/Integration/romberg.m
function [I] = romberg(f, a, b, nl)
%romberg Numerical Integration of functions using Romberg's method
% I = romberg(f,a,b,nl) Calculates the integral of the function f
% between the limits [a b]. It does so using nl initial estimates of
% the integral using the trapezoid method with successively smaller widths.
% 
% NOTE: f should be a vectorized function such that it can carry out
% element wise multiplication and division on a vector.
%
% NOTE2: nl > 25 should be avoided as they take a _long_ time.

  % Error checks
  if nl <= 0 || mod(nl, 1) ~= 0
    error('Integration level must be a positive integer');
  end

  intTable = nan(nl); % Empty matrix to use as as a divided difference
                      % table
  
  % Calculate the first column of the diff table using the compound trapezium 
  % rule with successively smaller intervals.
  h = (b - a);
  for k = 1:nl
    intTable(k,1) = 0.5 * h *(f(a) + 2*sum(f(a+h:h:b-h)) + f(b));
    h = h / 2;
  end

  % Calculate the remaining elements in the divided difference table using
  % Romberg's method.
  for kk = 2:nl
    for jj = length(intTable)-kk+1:-1:1
      intTable(jj, kk) = ((4^(kk-1) * intTable(jj+1, kk-1)) - intTable(jj, kk-1))/((4^(kk-1)) - 1);
    end
  end
  I = intTable(1,end);
end

File: Julia/Integration/romberg.jl
function romberg(f::Function, a::Number, b::Number, nl::Int)
  #= Numerical Integration of functions using Romberg's method.
  Keyword arguments:
  f -- Anonymous functon to integrate
  a -- Lower limit of integration
  b -- Upper limit of integration
  nl -- The level of integration to go to.

  Returns:
  The result of the integration of f between a & b
  =#

  intTable = zeros(nl, nl); # Empty matrix to use as div diff table

  # Calculate the first column of the div diff table using the trapezium rule.
  h = (b - a);
  for k = 1:nl
    intTable[k,1] = 0.5 * h * (f(a) + 2*sum(f([a+h:h:b-h])) + f(b));
    h /= 2;
  end

  # Calculate the remaining elements using Romberg's method.
  for k = 2:nl, j = nl-k+1:-1:1
    intTable[j, k] = ((4^(k-1))intTable[j+1, k-1] - intTable[j, k-1])/((4^(k-1)) - 1);    
  end

  return intTable[1, end];
end

function romberg(intEstimates::Array{Float64,1})
  #= Estiamte integrals from sets of data using the Romberg Method.

  Keyword arguments:
  intEstimates -- An array of initial estimates of integrals

  Returns:
    A tuple containing the integral and the relative error in the estimate 
  =# 
  intTable = zeros(length(intEstimates), length(intEstimates));
  relativeErrorTable = zeros(size(intEstimates));

  intTable[:, 1] = intEstimates;

  for k = 2:size(intTable)[1]
    for j = size(intTable)[1]-k+1:-1:1
      intTable[j, k] = ((4^(k-1))intTable[j+1, k-1] - intTable[j, k-1])/((4^(k-1)) - 1);
    end
  end

  for k = 2:size(intTable)[1]
    relativeErrorTable[k, 1] = 100 * abs((intTable[1, k] - intTable[2, k-1]) / intTable[1, k]);
  end

  return (intTable[1, end], relativeErrorTable[1, end]);
end
